lib/logstash/filters/kv.rb

```diff
@@ -238,27 +239,39 @@ def filter(event)
     # Add default key-values for missing keys
     kv = @default_keys.merge(kv)
 
-    # If we have any keys, create/append the hash
-    if kv.length > 0
-      if @target.nil?
-        # Default is to write to the root of the event.
-        dest = event.to_hash
-      else
-        if !event[@target].is_a?(Hash)
-          @logger.debug("Overwriting existing target field", :target => @target)
-          dest = event[@target] = {}
-        else
-          dest = event[@target]
-        end
+    unless kv.empty?
+      if @target
+        event.remove("#{@target}")
+        @logger.debug? && @logger.debug("Overwriting existing target field", :target => @target)
       end
 
-      dest.merge!(kv)
+      inject_fields(event, @target, kv)
+
       filter_matched(event)
     end
   end
 
   private
 
+  # recursiverly inject the data Hash content into event. the data parameter must be Hash on initial call.
+  #
+  # this is important to do here because we cannot assume in the future the in-place mutability of the Event#to_hash structure.
+  # TODO: (colin) we can certainly optimize this by introducing a new Event API to directly set fields without going through the
```

Comment:
ok, so it's not different for when set a hash to a field reference path but there is no way to set a hash at the root, i.e. when `@target` is not set.

maybe this could be optimized to only handle the root case and avoid injecting when in a path. good observation. 

also, if we agree in elastic/logstash#4293 on my suggestion for a stop-gap special field ref to set the root data, we will be able to completely remove this inject thing here. would love to have you feedback in elastic/logstash#4293